File: 1101-the-earliest-moment-when-everyone-become-friends/1101-the-earliest-moment-when-everyone-become-friends.cpp
struct DSU {
public:
    int components;
    vector<int> Parent, Rank;

    DSU(int n) : components(n), Parent(n), Rank(n, 1) { iota(Parent.begin(), Parent.end(), 0); }

    int find(int x) {
        while (x != Parent[x]) x = Parent[x] = Parent[Parent[x]];
        return x;
    }

    bool merge(int x, int y) {
        x = find(x);
        y = find(y);
        if (x == y) return false;
        if (Rank[x] > Rank[y]) swap(x, y);
        Parent[x] = y;
        Rank[y] += Rank[x];
        --components;
        return true;
    }
};

class Solution {
public:
    int earliestAcq(vector<vector<int>>& logs, int n) {
        DSU dsu(n);
        sort(logs.begin(), logs.end(), [&](const auto & A, const auto & B) {return A[0] < B[0];});
        for (auto &log : logs) {
            dsu.merge(log[1], log[2]);
            if (dsu.components == 1) return log[0];
        }
        return (dsu.components == 1) ? logs.back()[0] : -1;
    }
};

File: 1101-the-earliest-moment-when-everyone-become-friends/1101-the-earliest-moment-when-everyone-become-friends.java
class DSU {
    int n, components;
    int[] parent;
    int[] rank;
    DSU(int n) {
        this.n = n;
        this.components = n;
        this.rank = new int[n];
        this.parent = new int[n];
        for (int i = 0; i < n; ++i) parent[i] = i;
    }
    public int find(int x) {
        while (x != parent[x]) x = parent[x] = parent[parent[x]];
        return x;
    }
    public boolean merge(int x, int y) {
        x = find(x);
        y = find(y);
        if (x == y) return false;
        if (rank[x] > rank[y]) {int z = y; y = x; x = z;}
        parent[x] = y;
        rank[y] += rank[x];
        --components;
        return true;
    }
    public int get_components() { return this.components; }
    public boolean is_same_component(int x, int y) {return (find(x) == find(y));}
}

class Solution {
    public int earliestAcq(int[][] logs, int n) {
        Arrays.sort(logs, (log1, log2) -> log1[0] - log2[0]);
        DSU dsu = new DSU(n);
        int ans = 0;
        for (int i = 0; i < logs.length; ++i) {
            if (dsu.merge(logs[i][1], logs[i][2])) {
                ans = Math.max(ans, logs[i][0]);
            }
        }
        return ((dsu.get_components() == 1) ? ans : -1);
    }
}